src/inmanta/data/__init__.py

```diff
@@ -4031,6 +4093,153 @@ def to_dict(self) -> JsonType:
+    """ Represents the ordering by which resource logs should be sorted"""
+
+    validator_dataclass = ResourceAction
+    valid_sort_columns = {"timestamp": datetime.datetime}
+
+    @classmethod
+    def parse_from_string(
+        cls,
+        sort: str,
+    ) -> "ResourceLogOrder":
+        database_order = super().parse_from_string(sort)
+        return ResourceLogOrder(database_order.order_by_column, database_order.order)
+
+
+class CompileReportOrder(DatabaseOrder):
+    """ Represents the ordering by which compile reports should be sorted"""
+
+    validator_dataclass = Compile
+    valid_sort_columns = {"requested": datetime.datetime}
+
+    @classmethod
+    def parse_from_string(
+        cls,
+        sort: str,
+    ) -> "CompileReportOrder":
+        database_order = super().parse_from_string(sort)
+        return CompileReportOrder(database_order.order_by_column, database_order.order)
+
+
+class QueryBuilder:
+    def __init__(self, base_query: str, *values: List[object], initial_filter_statements: Optional[List[str]] = None) -> None:
+        self.sql_query = base_query
+        self.values = list(values)
+        if initial_filter_statements:
+            self.sql_query += BaseDocument._join_filter_statements(initial_filter_statements)
+        self.filter_started = bool(initial_filter_statements)
+
+    @property
+    def offset(self):
+        return len(self.values) + 1
+
+    def filter(self, filter_statements: List[str], values: List[object]) -> "QueryBuilder":
+        if self.filter_started:
+            self.sql_query += BaseDocument._continue_filter_statements(filter_statements)
+        else:
+            self.sql_query += BaseDocument._join_filter_statements(filter_statements)
+            self.filter_started = True
+        self.values.extend(values)
+        return self
+
+    def order_and_limit(
+        self, db_order: DatabaseOrder, limit: Optional[int] = None, backward_paging: Optional[bool] = False
+    ) -> "QueryBuilder":
+        order = db_order.get_order()
+        order_by_column = db_order.get_order_by_db_column_name()
+
+        if backward_paging:
+            backward_paging_order = order.invert().name
+
+            self.sql_query += f" ORDER BY {order_by_column} {backward_paging_order}, id {backward_paging_order}"
+        else:
+            self.sql_query += f" ORDER BY {order_by_column} {order}, id {order}"
+        if limit is not None:
+            if limit > DBLIMIT:
+                raise InvalidQueryParameter(f"Limit cannot be bigger than {DBLIMIT}, got {limit}")
+            elif limit > 0:
+                self.sql_query += " LIMIT " + str(limit)
+
+        if backward_paging:
+            self.sql_query = f"""SELECT * FROM ({self.sql_query}) AS matching_records
+                                ORDER BY matching_records.{order_by_column} {order},
+                                         matching_records.{db_order.id_column} {order}"""
+
+        return self
+
+    def page_count(
```

Comment:
This method here will only produce correct results if the base query has a very particular form (i.e.`SELECT ...` is missing. )

This kind of defeats the purpose of a query builder as an abstraction to handle the complexity of building queries.

I would split the query builder in two types:

- OpaqueQueryBuilder, which works like this one
- A QueryBuilder that is a subclass, that keeps all the parts of the query separated (projection, filter, order, table)

The QueryBuilder can support page_count, the OpaqueQueryBuilder can not.
complex things like order_and_limit always return OpaqueQueryBuilder

I'm going to ask @bartv for a second opinion on this.

Reply:
Arent we getting to the point that a class like this should symbolicly build the actual query and then a render method that actuall generates the query? I have the feeling a lot of complexity comes from the fact that each method adds to a string which is a irreversible operation.

But not entirely sure, I do not have enough feeling with the code yet.

Reply:
I think we are thinking along similar lines